{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this script we process the raw data coming from the preprocessing to standarize the data into a working Anndata file for the rest of the analysis.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as scp\n",
    "import scipy.sparse as sprs\n",
    "import pandas as pd\n",
    "import warnings\n",
    "import os\n",
    "import numpy as np\n",
    "\n",
    "warnings.filterwarnings(\"ignore\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Gastruloid\n",
    "\n",
    "In this step, we create the dataset for the AnnData format. \n",
    "\n",
    "We incorporate to the data the information coming from:\n",
    "\n",
    "1. The raw count matrix.\n",
    "2. The information coming from the flow cytometer sorting procedure."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "genes = pd.read_csv(\"Data/Genes.csv\")\n",
    "cells = pd.read_csv(\"Data/Cells.csv\")\n",
    "m = sprs.load_npz(\"Data/count_matrix_total.npz\")\n",
    "\n",
    "#Adding plate information\n",
    "#cells[\"Plate\"] = \n",
    "\n",
    "statistics = pd.DataFrame()\n",
    "statistics.loc[0,[\"Plates\",\"Wells per plate\",\"Total Wells\"]] = [9,96,9*96]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "18.0"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(1715+13)/96"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>...</th>\n",
       "      <th>53690</th>\n",
       "      <th>53691</th>\n",
       "      <th>53692</th>\n",
       "      <th>53693</th>\n",
       "      <th>53694</th>\n",
       "      <th>53695</th>\n",
       "      <th>53696</th>\n",
       "      <th>53697</th>\n",
       "      <th>53698</th>\n",
       "      <th>53699</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>235</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>82</td>\n",
       "      <td>14</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>272</td>\n",
       "      <td>0</td>\n",
       "      <td>102</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>314</td>\n",
       "      <td>0</td>\n",
       "      <td>138</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>236</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>98</td>\n",
       "      <td>102</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>494</td>\n",
       "      <td>0</td>\n",
       "      <td>134</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>64</td>\n",
       "      <td>12</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>226</td>\n",
       "      <td>0</td>\n",
       "      <td>372</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>176</td>\n",
       "      <td>0</td>\n",
       "      <td>60</td>\n",
       "      <td>72</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1710</th>\n",
       "      <td>142</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>107</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1711</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1712</th>\n",
       "      <td>44</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>136</td>\n",
       "      <td>32</td>\n",
       "      <td>72</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1713</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1714</th>\n",
       "      <td>294</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1715 rows Ã— 53700 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      0      1      2      3      4      5      6      7      8      9      \\\n",
       "0       235      0     11      0      0      0      0      0     82     14   \n",
       "1       272      0    102      0      0      0    314      0    138      0   \n",
       "2       236      0      8      0      0      0      0      0     98    102   \n",
       "3       494      0    134      0      0      0      0      0     64     12   \n",
       "4       226      0    372      0      0      0    176      0     60     72   \n",
       "...     ...    ...    ...    ...    ...    ...    ...    ...    ...    ...   \n",
       "1710    142      0      0      0      0      0      0      0    107      0   \n",
       "1711      0      0      0      0      0      0      0      0      0      0   \n",
       "1712     44      0      0      0      0      0    136     32     72      0   \n",
       "1713      0      0      0      0      0      0      0      0      0      0   \n",
       "1714    294      0      0      0      0      0      8      0      0      0   \n",
       "\n",
       "      ...  53690  53691  53692  53693  53694  53695  53696  53697  53698  \\\n",
       "0     ...      0      0      0      0      0      0      0      0      0   \n",
       "1     ...      0      0      0      0      0      0      0      0      0   \n",
       "2     ...      0      0      0      0      0      0      0      0      0   \n",
       "3     ...      0      0      0      0      0      0      0      0      0   \n",
       "4     ...      0      0      0      0      0      0      0      2      0   \n",
       "...   ...    ...    ...    ...    ...    ...    ...    ...    ...    ...   \n",
       "1710  ...      0      0      0      0      0      0      0      0      0   \n",
       "1711  ...      0      0      0      0      0      0      0      0      0   \n",
       "1712  ...      0      0      0      0      0      0      0      0      0   \n",
       "1713  ...      0      0      0      0      0      0      0      0      0   \n",
       "1714  ...      0      0      0      0      0      0      0      0      0   \n",
       "\n",
       "      53699  \n",
       "0         0  \n",
       "1         0  \n",
       "2         0  \n",
       "3         0  \n",
       "4         0  \n",
       "...     ...  \n",
       "1710      0  \n",
       "1711      0  \n",
       "1712      0  \n",
       "1713      0  \n",
       "1714      0  \n",
       "\n",
       "[1715 rows x 53700 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.DataFrame(np.array(m.todense()))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Merging lanes <a name=\"Merging\"></a>\n",
    "\n",
    "In the preprocessing step, we aligned the files taking the different lanes as separated measures, however, these measures corrrespond to the cells. In this step, we unified the measures coming from different lanes measures coming from a unique well. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HG5V7DRXY_2_IDT-DUI-NXT-341  at  216hr _ CD45  doesn't exist. We only have one lane measure of this well.\n"
     ]
    }
   ],
   "source": [
    "dropIndex = [] # For dropping the index that are merged\n",
    "statistics.loc[0,\"singleLane\"] = 0\n",
    "for j,i in zip(cells[cells[\"Lane\"]==1].index,cells[cells[\"Lane\"]==1].values):\n",
    "    \n",
    "    # making the conditions to univocaly finding the cell with the other corresponding lane\n",
    "    name = i[0]\n",
    "    time = i[2]\n",
    "    condition = i[3]\n",
    "    \n",
    "    cond1 = cells[\"Cell\"] == name.replace(\"_1_\",\"_2_\")\n",
    "    cond2 = cells[\"Time\"] == time\n",
    "    cond3 = cells[\"Lane\"] == 2\n",
    "\n",
    "    cell = cells[cond1.values*cond2.values*cond3.values]\n",
    "\n",
    "    #Add the\n",
    "    if cell.shape[0] == 1:\n",
    "        m[j,:] += m[cell.index,:]\n",
    "        dropIndex.append(cell.index[0])\n",
    "    else:\n",
    "        print(name.replace(\"_1_\",\"_2_\"), \" at \", time,\"_\",condition, \" doesn't exist. We only have one lane measure of this well.\")\n",
    "        statistics.loc[0,\"singleLane\"] += 1\n",
    "\n",
    "#Drop merged cells\n",
    "cells = cells.drop(dropIndex)\n",
    "\n",
    "#Print cells that only have measures in second lane\n",
    "for j,i in zip(cells[cells[\"Lane\"]==2].index,cells[cells[\"Lane\"]==2].values):\n",
    "    name = i[0]\n",
    "    time = i[1]\n",
    "    condition = i[2]\n",
    "    print(name.replace(\"_2_\",\"_1_\"), \" at \", time,\"_\",condition, \" doesn't exist. We only have one lane measure of this well.\")\n",
    "    statistics.loc[0,\"singleLane\"] += 1\n",
    "#Change name of cells that are merged\n",
    "cells[\"Cell\"] = [i.replace(\"_1_\",\"_\").replace(\"_2_\",\"_\") for i in cells[\"Cell\"].values]\n",
    "\n",
    "#Retain only rows that correspond to merged cells\n",
    "m = m[cells.index]\n",
    "\n",
    "#Reset index of the cells\n",
    "cells.reset_index(drop=True,inplace=True)\n",
    "\n",
    "#Drop Lane information\n",
    "cells.drop(columns=\"Lane\",inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Standarize the tags <a name=\"Standarize\"></a>\n",
    "\n",
    "We can check that in the raw data some tags have the same meaning but different nomenclature because of how they where given to the sequencing facility. We standarize the nomenclature.\n",
    "\n",
    "1. kitScal and kitscal in Condition\n",
    "2. empty and EMPTY in BiologicalSample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Plate :  ['P1' 'P6' 'P7' 'P8' 'P9' 'P2' 'P4' 'P3' 'P5']\n",
      "Time :  ['120hr' '192hr' '216hr' '144hr' '168hr']\n",
      "Condition :  ['whole' 'CD45' 'kitscal' 'kitScal' 'CD41']\n",
      "BiologicalSample :  ['P1' 'P2' 'empty' 'EMPTY' 'black']\n"
     ]
    }
   ],
   "source": [
    "for i in cells.columns[1:]:\n",
    "    print(i,\": \", cells[i].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "cells.loc[cells[\"Condition\"]==\"kitscal\",\"Condition\"] = \"kitScal\"\n",
    "cells.loc[cells[\"BiologicalSample\"]==\"EMPTY\",\"BiologicalSample\"] = \"empty\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Additionally, we have a few cells marked with the tag black. Those correspond to measures did no have a valid measurement in the flow cytommetry system. This correspond to confirmed doublets by the flow cytometry sorting process as can be seen in the next section."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Cell</th>\n",
       "      <th>Plate</th>\n",
       "      <th>Time</th>\n",
       "      <th>Condition</th>\n",
       "      <th>BiologicalSample</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>583</th>\n",
       "      <td>HG72GDRXY_IDT-DUI-NXT-198</td>\n",
       "      <td>P3</td>\n",
       "      <td>144hr</td>\n",
       "      <td>CD41</td>\n",
       "      <td>black</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>598</th>\n",
       "      <td>HG72GDRXY_IDT-DUI-NXT-210</td>\n",
       "      <td>P3</td>\n",
       "      <td>144hr</td>\n",
       "      <td>CD41</td>\n",
       "      <td>black</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>611</th>\n",
       "      <td>HG72GDRXY_IDT-DUI-NXT-222</td>\n",
       "      <td>P3</td>\n",
       "      <td>144hr</td>\n",
       "      <td>CD41</td>\n",
       "      <td>black</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          Cell Plate   Time Condition BiologicalSample\n",
       "583  HG72GDRXY_IDT-DUI-NXT-198    P3  144hr      CD41            black\n",
       "598  HG72GDRXY_IDT-DUI-NXT-210    P3  144hr      CD41            black\n",
       "611  HG72GDRXY_IDT-DUI-NXT-222    P3  144hr      CD41            black"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cells[cells[\"BiologicalSample\"]==\"black\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Adding Flow cytommetry measures to data <a name=\"Add_flow\"></a>\n",
    "\n",
    "For this experimental setup, the data has been sorted to enrich for rare cells. We add the measures of the cells to the available information of the system.\n",
    "\n",
    "We check that there are wells that have flow information but there are not corresponding RNA sequenced measures.\n",
    "\n",
    "Additionally we confirm that some wells were measured more than once indicating that the well contains doublets, this where marked already by the tag **black** in biological condition."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "P2_144hr_whole   A11  does not have scRNA data. Reading error.\n",
      "P9_216hr_CD45   B5  does not have scRNA data. Reading error.\n",
      "P9_216hr_CD45   C5  does not have scRNA data. Reading error.\n",
      "P9_216hr_CD45   D5  does not have scRNA data. Reading error.\n",
      "P9_216hr_CD45   F5  does not have scRNA data. Reading error.\n",
      "P9_216hr_CD45   H5  does not have scRNA data. Reading error.\n",
      "583 HG72GDRXY_IDT-DUI-NXT-198  is measured twice. Doublet.\n",
      "598 HG72GDRXY_IDT-DUI-NXT-210  is measured twice. Doublet.\n",
      "611 HG72GDRXY_IDT-DUI-NXT-222  is measured twice. Doublet.\n"
     ]
    }
   ],
   "source": [
    "# List the sorting files by plates\n",
    "plates = os.listdir(\"Data/SMARTseq\")\n",
    "#Add identifier flow cytometry file\n",
    "cells[\"Flow Cytommetry File\"] = np.NaN\n",
    "\n",
    "#Counters\n",
    "statistics.loc[0,\"doublet\"] = 0\n",
    "statistics.loc[0,\"noRNA\"] = 0\n",
    "\n",
    "for plate in plates:\n",
    "\n",
    "    #Load table mapping well indexes to cell indexes\n",
    "    table = pd.read_excel(\"Data/Mapping.xlsx\",sheet_name=plate).set_index(\"X1\").unstack()\n",
    "\n",
    "    #Make mapping list of cells to plates\n",
    "    mapping = {i[1]+str(i[0]):str(j).replace(\"_1_\",\"_\") for i,j in zip(table.index, table.values)}\n",
    "\n",
    "    #List and go over the sorting files splited by plate\n",
    "    for file_num,file_name in enumerate(os.listdir(\"Data/SMARTseq/\"+plate)):\n",
    "\n",
    "        #Load table of measures for the plate and\n",
    "        data = pd.read_csv(\"Data/SMARTseq/\"+plate+\"/\"+file_name,skiprows=9,decimal=\",\")\n",
    "        data.set_index(\"Well\",inplace=True)\n",
    "\n",
    "        #Add information to the cell\n",
    "        parameters = [i for i in data.columns if \"All\" in i and \"Median\" in i and \"Time\" not in i]\n",
    "        parameters_clean = [i.split(\"Events \")[1].split(\" Median\")[0] for i in parameters]\n",
    "\n",
    "        for i in data.index:\n",
    "                if cells[cells[\"Cell\"] == mapping[i]].shape[0] > 0:\n",
    "                    if type(cells.loc[cells[\"Cell\"] == mapping[i],\"Flow Cytommetry File\"].values[0]) == str:\n",
    "                        print(cells.loc[cells[\"Cell\"] == mapping[i]].index[0],cells.loc[cells[\"Cell\"] == mapping[i],\"Cell\"].values[0], \" is measured twice. Doublet.\")\n",
    "                        statistics.loc[0,\"doublet\"] += 1\n",
    "                    else:\n",
    "                        cells.loc[cells[\"Cell\"] == mapping[i],\"Flow Cytommetry File\"] = plate+\"_\"+str(file_num)\n",
    "                        for k,j in enumerate(parameters):\n",
    "                            cells.loc[cells[\"Cell\"] == mapping[i],parameters_clean[k]] = data.loc[i,j]\n",
    "                else:\n",
    "                    print(plate,\" \",i, \" does not have scRNA data. Reading error.\")\n",
    "                    statistics.loc[0,\"noRNA\"] += 1\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Finally, we check for wells that have transcription bu do not have any flow cytometry measure. Consistently, most of them correspond to wells called as empty, except for two samples that should be filled. These wells are wells that did not have any cell deposted in them."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Cell</th>\n",
       "      <th>Plate</th>\n",
       "      <th>Time</th>\n",
       "      <th>Condition</th>\n",
       "      <th>BiologicalSample</th>\n",
       "      <th>Flow Cytommetry File</th>\n",
       "      <th>FSC-A</th>\n",
       "      <th>FSC-H</th>\n",
       "      <th>FSC-W</th>\n",
       "      <th>SSC-A</th>\n",
       "      <th>SSC-H</th>\n",
       "      <th>SSC-W</th>\n",
       "      <th>GFP FLK-1 530/30 Blue-A</th>\n",
       "      <th>582/15 Y/G-A</th>\n",
       "      <th>PE Dazzle 594 CD41 610/20 Y/G-A</th>\n",
       "      <th>PECy7 Sca-1 780/60 Y/G-A</th>\n",
       "      <th>APC CD45 660/20 Red-A</th>\n",
       "      <th>APCCy7 cKIT 780/60 Red-A</th>\n",
       "      <th>Hoechst 58 450/40 Violet-A</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>HG3JFDRXY_IDT-DUI-NXT-282</td>\n",
       "      <td>P1</td>\n",
       "      <td>120hr</td>\n",
       "      <td>whole</td>\n",
       "      <td>empty</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>HG3JFDRXY_IDT-DUI-NXT-288</td>\n",
       "      <td>P1</td>\n",
       "      <td>120hr</td>\n",
       "      <td>whole</td>\n",
       "      <td>empty</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>192</th>\n",
       "      <td>HG5V7DRXY_IDT-DUI-NXT-186</td>\n",
       "      <td>P7</td>\n",
       "      <td>192hr</td>\n",
       "      <td>CD45</td>\n",
       "      <td>empty</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>199</th>\n",
       "      <td>HG5V7DRXY_IDT-DUI-NXT-192</td>\n",
       "      <td>P7</td>\n",
       "      <td>192hr</td>\n",
       "      <td>CD45</td>\n",
       "      <td>empty</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>299</th>\n",
       "      <td>HG5V7DRXY_IDT-DUI-NXT-282</td>\n",
       "      <td>P8</td>\n",
       "      <td>192hr</td>\n",
       "      <td>kitScal</td>\n",
       "      <td>empty</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>305</th>\n",
       "      <td>HG5V7DRXY_IDT-DUI-NXT-288</td>\n",
       "      <td>P8</td>\n",
       "      <td>192hr</td>\n",
       "      <td>kitScal</td>\n",
       "      <td>empty</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>400</th>\n",
       "      <td>HG5V7DRXY_IDT-DUI-NXT-378</td>\n",
       "      <td>P9</td>\n",
       "      <td>216hr</td>\n",
       "      <td>CD45</td>\n",
       "      <td>empty</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>407</th>\n",
       "      <td>HG5V7DRXY_IDT-DUI-NXT-384</td>\n",
       "      <td>P9</td>\n",
       "      <td>216hr</td>\n",
       "      <td>CD45</td>\n",
       "      <td>empty</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>465</th>\n",
       "      <td>HG5V7DRXY_IDT-DUI-NXT-90</td>\n",
       "      <td>P6</td>\n",
       "      <td>192hr</td>\n",
       "      <td>whole</td>\n",
       "      <td>empty</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>471</th>\n",
       "      <td>HG5V7DRXY_IDT-DUI-NXT-96</td>\n",
       "      <td>P6</td>\n",
       "      <td>192hr</td>\n",
       "      <td>whole</td>\n",
       "      <td>empty</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>570</th>\n",
       "      <td>HG72GDRXY_IDT-DUI-NXT-186</td>\n",
       "      <td>P4</td>\n",
       "      <td>144hr</td>\n",
       "      <td>kitScal</td>\n",
       "      <td>empty</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>577</th>\n",
       "      <td>HG72GDRXY_IDT-DUI-NXT-192</td>\n",
       "      <td>P4</td>\n",
       "      <td>144hr</td>\n",
       "      <td>kitScal</td>\n",
       "      <td>empty</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>625</th>\n",
       "      <td>HG72GDRXY_IDT-DUI-NXT-235</td>\n",
       "      <td>P3</td>\n",
       "      <td>144hr</td>\n",
       "      <td>CD41</td>\n",
       "      <td>P2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>677</th>\n",
       "      <td>HG72GDRXY_IDT-DUI-NXT-282</td>\n",
       "      <td>P3</td>\n",
       "      <td>144hr</td>\n",
       "      <td>CD41</td>\n",
       "      <td>empty</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>683</th>\n",
       "      <td>HG72GDRXY_IDT-DUI-NXT-288</td>\n",
       "      <td>P3</td>\n",
       "      <td>144hr</td>\n",
       "      <td>CD41</td>\n",
       "      <td>empty</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>684</th>\n",
       "      <td>HG72GDRXY_IDT-DUI-NXT-289</td>\n",
       "      <td>P5</td>\n",
       "      <td>168hr</td>\n",
       "      <td>whole</td>\n",
       "      <td>P1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>790</th>\n",
       "      <td>HG72GDRXY_IDT-DUI-NXT-384</td>\n",
       "      <td>P5</td>\n",
       "      <td>168hr</td>\n",
       "      <td>whole</td>\n",
       "      <td>empty</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>848</th>\n",
       "      <td>HG72GDRXY_IDT-DUI-NXT-90</td>\n",
       "      <td>P2</td>\n",
       "      <td>144hr</td>\n",
       "      <td>whole</td>\n",
       "      <td>empty</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>854</th>\n",
       "      <td>HG72GDRXY_IDT-DUI-NXT-96</td>\n",
       "      <td>P2</td>\n",
       "      <td>144hr</td>\n",
       "      <td>whole</td>\n",
       "      <td>empty</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          Cell Plate   Time Condition BiologicalSample  \\\n",
       "89   HG3JFDRXY_IDT-DUI-NXT-282    P1  120hr     whole            empty   \n",
       "95   HG3JFDRXY_IDT-DUI-NXT-288    P1  120hr     whole            empty   \n",
       "192  HG5V7DRXY_IDT-DUI-NXT-186    P7  192hr      CD45            empty   \n",
       "199  HG5V7DRXY_IDT-DUI-NXT-192    P7  192hr      CD45            empty   \n",
       "299  HG5V7DRXY_IDT-DUI-NXT-282    P8  192hr   kitScal            empty   \n",
       "305  HG5V7DRXY_IDT-DUI-NXT-288    P8  192hr   kitScal            empty   \n",
       "400  HG5V7DRXY_IDT-DUI-NXT-378    P9  216hr      CD45            empty   \n",
       "407  HG5V7DRXY_IDT-DUI-NXT-384    P9  216hr      CD45            empty   \n",
       "465   HG5V7DRXY_IDT-DUI-NXT-90    P6  192hr     whole            empty   \n",
       "471   HG5V7DRXY_IDT-DUI-NXT-96    P6  192hr     whole            empty   \n",
       "570  HG72GDRXY_IDT-DUI-NXT-186    P4  144hr   kitScal            empty   \n",
       "577  HG72GDRXY_IDT-DUI-NXT-192    P4  144hr   kitScal            empty   \n",
       "625  HG72GDRXY_IDT-DUI-NXT-235    P3  144hr      CD41               P2   \n",
       "677  HG72GDRXY_IDT-DUI-NXT-282    P3  144hr      CD41            empty   \n",
       "683  HG72GDRXY_IDT-DUI-NXT-288    P3  144hr      CD41            empty   \n",
       "684  HG72GDRXY_IDT-DUI-NXT-289    P5  168hr     whole               P1   \n",
       "790  HG72GDRXY_IDT-DUI-NXT-384    P5  168hr     whole            empty   \n",
       "848   HG72GDRXY_IDT-DUI-NXT-90    P2  144hr     whole            empty   \n",
       "854   HG72GDRXY_IDT-DUI-NXT-96    P2  144hr     whole            empty   \n",
       "\n",
       "    Flow Cytommetry File  FSC-A  FSC-H  FSC-W  SSC-A  SSC-H  SSC-W  \\\n",
       "89                   NaN    NaN    NaN    NaN    NaN    NaN    NaN   \n",
       "95                   NaN    NaN    NaN    NaN    NaN    NaN    NaN   \n",
       "192                  NaN    NaN    NaN    NaN    NaN    NaN    NaN   \n",
       "199                  NaN    NaN    NaN    NaN    NaN    NaN    NaN   \n",
       "299                  NaN    NaN    NaN    NaN    NaN    NaN    NaN   \n",
       "305                  NaN    NaN    NaN    NaN    NaN    NaN    NaN   \n",
       "400                  NaN    NaN    NaN    NaN    NaN    NaN    NaN   \n",
       "407                  NaN    NaN    NaN    NaN    NaN    NaN    NaN   \n",
       "465                  NaN    NaN    NaN    NaN    NaN    NaN    NaN   \n",
       "471                  NaN    NaN    NaN    NaN    NaN    NaN    NaN   \n",
       "570                  NaN    NaN    NaN    NaN    NaN    NaN    NaN   \n",
       "577                  NaN    NaN    NaN    NaN    NaN    NaN    NaN   \n",
       "625                  NaN    NaN    NaN    NaN    NaN    NaN    NaN   \n",
       "677                  NaN    NaN    NaN    NaN    NaN    NaN    NaN   \n",
       "683                  NaN    NaN    NaN    NaN    NaN    NaN    NaN   \n",
       "684                  NaN    NaN    NaN    NaN    NaN    NaN    NaN   \n",
       "790                  NaN    NaN    NaN    NaN    NaN    NaN    NaN   \n",
       "848                  NaN    NaN    NaN    NaN    NaN    NaN    NaN   \n",
       "854                  NaN    NaN    NaN    NaN    NaN    NaN    NaN   \n",
       "\n",
       "     GFP FLK-1 530/30 Blue-A  582/15 Y/G-A  PE Dazzle 594 CD41 610/20 Y/G-A  \\\n",
       "89                       NaN           NaN                              NaN   \n",
       "95                       NaN           NaN                              NaN   \n",
       "192                      NaN           NaN                              NaN   \n",
       "199                      NaN           NaN                              NaN   \n",
       "299                      NaN           NaN                              NaN   \n",
       "305                      NaN           NaN                              NaN   \n",
       "400                      NaN           NaN                              NaN   \n",
       "407                      NaN           NaN                              NaN   \n",
       "465                      NaN           NaN                              NaN   \n",
       "471                      NaN           NaN                              NaN   \n",
       "570                      NaN           NaN                              NaN   \n",
       "577                      NaN           NaN                              NaN   \n",
       "625                      NaN           NaN                              NaN   \n",
       "677                      NaN           NaN                              NaN   \n",
       "683                      NaN           NaN                              NaN   \n",
       "684                      NaN           NaN                              NaN   \n",
       "790                      NaN           NaN                              NaN   \n",
       "848                      NaN           NaN                              NaN   \n",
       "854                      NaN           NaN                              NaN   \n",
       "\n",
       "     PECy7 Sca-1 780/60 Y/G-A  APC CD45 660/20 Red-A  \\\n",
       "89                        NaN                    NaN   \n",
       "95                        NaN                    NaN   \n",
       "192                       NaN                    NaN   \n",
       "199                       NaN                    NaN   \n",
       "299                       NaN                    NaN   \n",
       "305                       NaN                    NaN   \n",
       "400                       NaN                    NaN   \n",
       "407                       NaN                    NaN   \n",
       "465                       NaN                    NaN   \n",
       "471                       NaN                    NaN   \n",
       "570                       NaN                    NaN   \n",
       "577                       NaN                    NaN   \n",
       "625                       NaN                    NaN   \n",
       "677                       NaN                    NaN   \n",
       "683                       NaN                    NaN   \n",
       "684                       NaN                    NaN   \n",
       "790                       NaN                    NaN   \n",
       "848                       NaN                    NaN   \n",
       "854                       NaN                    NaN   \n",
       "\n",
       "     APCCy7 cKIT 780/60 Red-A  Hoechst 58 450/40 Violet-A  \n",
       "89                        NaN                         NaN  \n",
       "95                        NaN                         NaN  \n",
       "192                       NaN                         NaN  \n",
       "199                       NaN                         NaN  \n",
       "299                       NaN                         NaN  \n",
       "305                       NaN                         NaN  \n",
       "400                       NaN                         NaN  \n",
       "407                       NaN                         NaN  \n",
       "465                       NaN                         NaN  \n",
       "471                       NaN                         NaN  \n",
       "570                       NaN                         NaN  \n",
       "577                       NaN                         NaN  \n",
       "625                       NaN                         NaN  \n",
       "677                       NaN                         NaN  \n",
       "683                       NaN                         NaN  \n",
       "684                       NaN                         NaN  \n",
       "790                       NaN                         NaN  \n",
       "848                       NaN                         NaN  \n",
       "854                       NaN                         NaN  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "statistics.loc[0,\"empty\"] = cells[cells[\"Flow Cytommetry File\"].isna()*cells[\"BiologicalSample\"]==\"empty\"].shape[0]\n",
    "statistics.loc[0,\"emptyWrong\"] = cells[cells[\"Flow Cytommetry File\"].isna()*np.array(cells[\"BiologicalSample\"]!=\"empty\")].shape[0]\n",
    "\n",
    "cells[cells[\"Flow Cytommetry File\"].isna()]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Statistics, create Annotated data file and save <a name=\"Statistics\"></a>\n",
    "\n",
    "We will save the data of the files with non-zero RNA sample measures but we will keep those measures that contain doublets and empty well for now so they appear in the QC check."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "percentage = 0\n",
    "percentage += statistics.loc[0,\"singleLane\"]\n",
    "percentage += statistics.loc[0,\"doublet\"]\n",
    "percentage += statistics.loc[0,\"noRNA\"]\n",
    "percentage += statistics.loc[0,\"noRNA\"]\n",
    "percentage += statistics.loc[0,\"noRNA\"]\n",
    "statistics.loc[0,\"wells with bad samples\"] = percentage\n",
    "statistics.loc[0,\"wells with good samples\"] = statistics.loc[0,\"Total Wells\"]-percentage\n",
    "statistics.loc[0,\"% wells with bad samples\"] = statistics.loc[0,\"wells with bad samples\"]*100/statistics.loc[0,\"Total Wells\"]\n",
    "statistics.loc[0,\"% wells with good samples\"] = statistics.loc[0,\"wells with good samples\"]*100/statistics.loc[0,\"Total Wells\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Plates</th>\n",
       "      <th>Wells per plate</th>\n",
       "      <th>Total Wells</th>\n",
       "      <th>singleLane</th>\n",
       "      <th>doublet</th>\n",
       "      <th>noRNA</th>\n",
       "      <th>empty</th>\n",
       "      <th>emptyWrong</th>\n",
       "      <th>wells with bad samples</th>\n",
       "      <th>wells with good samples</th>\n",
       "      <th>% wells with bad samples</th>\n",
       "      <th>% wells with good samples</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>9.0</td>\n",
       "      <td>96.0</td>\n",
       "      <td>864.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>842.0</td>\n",
       "      <td>2.546296</td>\n",
       "      <td>97.453704</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Plates  Wells per plate  Total Wells  singleLane  doublet  noRNA  empty  \\\n",
       "0     9.0             96.0        864.0         1.0      3.0    6.0   17.0   \n",
       "\n",
       "   emptyWrong  wells with bad samples  wells with good samples  \\\n",
       "0         2.0                    22.0                    842.0   \n",
       "\n",
       "   % wells with bad samples  % wells with good samples  \n",
       "0                  2.546296                  97.453704  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "statistics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Create the Annotated data\n",
    "adata = scp.AnnData(m,obs=cells,var=genes.set_index(\"gene_id\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "... storing 'Plate' as categorical\n",
      "... storing 'Time' as categorical\n",
      "... storing 'Condition' as categorical\n",
      "... storing 'BiologicalSample' as categorical\n",
      "... storing 'Flow Cytommetry File' as categorical\n",
      "... storing 'gene_name' as categorical\n"
     ]
    }
   ],
   "source": [
    "adata.write(\"Data/Raw.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "e7370f93d1d0cde622a1f8e1c04877d8463912d04d973331ad4851f04de6915a"
  },
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  },
  "toc-autonumbering": true,
  "toc-showcode": false,
  "toc-showmarkdowntxt": false,
  "toc-showtags": false
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
